### PR TITLE
adds support for domainMappings to serverless deployment

### PR DESCRIPTION
**Pull request by invincibleJai**
**Fixes**: 
https://issues.redhat.com/browse/ODC-5356

**Analysis / Root cause**: 
The user was not able to provide domain mapping for knative service from dev console

**Solution Description**: 
- Added ability to provide a custom domain for the routes configured for serverless applications for import flows/upload jar/knatify while creation and edit flows
- Visualisation of DomainMapping resources in KSVC sidebar in topology.
- Enabled dry run for knative serving resources

**Screen shots / Gifs for design review**: 

image

image

Kapture 2021-07-01 at 20 51 25

cc @openshift/team-devconsole-ux 

**Unit test coverage report**: 

- updated current tests

image

image

- [ ] add more tests

**Test setup:**

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by bgliwa01** [comment]
To be consistent with other areas of the console, change help text to "Add domain" please!

**Comment by invincibleJai** [comment]
> To be consistent with other areas of the console, change help text to "Add domain" please!

Thanks, have updated it 

image

**Comment by nak3** [comment]
Awesome! I like this so much. The Gif&Screenshot look good to me.

File: frontend/packages/dev-console/locales/en/devconsole.json [hunk 1]
```diff
@@ -417,6 +417,9 @@
   "A type of deployment that enables Serverless scaling to 0 when idle.": "A type of deployment that enables Serverless scaling to 0 when idle.",
   "Resources": "Resources",
   "Select the resource type to generate": "Select the resource type to generate",
+  "Domain Mapping": "Domain Mapping",
```
Comment (vikram-raj) [inline frontend/packages/dev-console/locales/en/devconsole.json:420]:
```suggestion
  "Domain Mapping": "Domain mapping",
```

File: frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts [hunk 2]
```diff
@@ -386,7 +389,7 @@ export const createOrUpdateDeployImageResources = async (
         requests.push(k8sCreate(RouteModel, route, dryRun ? dryRunOpt : {}));
       }
     }
-  } else if (!dryRun) {
+  } else if (formData.resources === Resources.KnativeService) {
     // Do not run serverless call during the dry run.
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts:393]: Nit: we can remove this comment now.

File: frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx [hunk 3]
```diff
@@ -1,27 +1,92 @@
...
+        label={t('devconsole~Target port')}
+        placeholderText={placeholderPort.toString()}
+        helpText={t('devconsole~Target port for traffic.')}
+        options={portOptions}
+      />
+      {domainMappingLoaded || domainMappingLoadErr ? (
+        <SelectInputField
+          data-test-id="kafkasource-bootstrapservers-field"
+          name="serverless.domainMapping"
+          label={t('devconsole~Domain Mapping')}
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx:78]:
```suggestion
          label={t('devconsole~Domain mapping')}
```

File: frontend/packages/knative-plugin/src/components/overview/KnativeServiceResources.tsx [hunk 4]
```diff
@@ -75,6 +83,12 @@ const KnativeServiceResources: React.FC<KnativeServiceResourceProps> = ({ item }
       {brokers.length > 0 && (
         <PubSubSubscribers subscribers={brokers} title={t('knative-plugin~Triggers')} />
       )}
+      {domainMappings.length > 0 && (
+        <DomainMappingOverviewList
+          domainMappings={domainMappings}
+          title={t('knative-plugin~Domain Mappings')}
```
Comment (vikram-raj) [inline frontend/packages/knative-plugin/src/components/overview/KnativeServiceResources.tsx:89]:
```suggestion
          title={t('knative-plugin~Domain mappings')}
```

File: frontend/packages/knative-plugin/src/utils/get-knative-resources.ts [hunk 5]
```diff
@@ -312,6 +339,19 @@ export const knativeCamelKameletBindingResourceWatchers = (
   };
 };
 
+export const knativeCamelDomainMappingResourceWatchers = (
+  namespace: string,
+): WatchK8sResources<any> => {
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/utils/get-knative-resources.ts:344]: nit:
```suggestion
): WatchK8sResources<{ [key: string]: K8sResourceKind[] }> => {
```

**Review by karthikjeeyar — APPROVED** [review]
Verified locally.
/lgtm

**Comment by invincibleJai** [comment]
/label px-approved
added on behalf of px as agreed upon.

/label docs-approved
added on behalf of the docs team as the epic has a doc-ack label.

**Review by vikram-raj — COMMENTED** [review]
Verified the changes and it works as expected.

/lgtm

**Comment by invincibleJai** [comment]
/hold 

needed QE-ACK

**Comment by invincibleJai** [comment]
/hold

added hold as seeing consistent failure in e2e tests

**Review by bgliwa01 — APPROVED** [review]
lgtm

**Pull request merged**